default/scripting/buildings/TERRAFORM.focs.txt

```diff
@@ -84,5 +84,80 @@ TARGET_TERRAFORMING
     icon = "icons/building/terraform.png"
 '''
 
+BuildingType
+    name = "BLD_TERRAFORM_BEST"
+    description = "BLD_TERRAFORM_BEST_DESC"
+    buildcost =  75 * (1 - 0.4 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_TERRAFORMING"])
+                    * Target.HabitableSize
+                      // calculating multiple steps: 1+2+..+N == (1+N)*N/2
+                    * max(1, (1 + (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType))
+                             * (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType) / 2 )
```

Comment:
That calculation can be quite wrong if the current type is not the original type. Especially it means that reverting to the original type costs one step, no matter how far it is away.

Reply:
reverting is not possible with that building

Reply:
but i certainly meant the current type, not the original one

Reply:
> reverting is not possible with that building

Of course it is, you just have to change the species to one that likes the original type. diff(Current, NextBest) is definitely better.

There is still the following issue, though: for 4 steps cost 10 x base cost, 3 steps cost 6, regardless of the distance form the original type. So if I want to terraform e.g. from Tundra to Swamp, I can use Terrafrom to Desert, followed by Popular Terraforming and its much cheaper then using the Popular Terraforming straight away.

Reply:
ok, i need to check how the costs are actually intended

Reply:
@Grummel7  could you help me figure out a correct formula? i am pretty much swamped. i guess in the old formula, each step cost was based on the distance to OriginalType. so the base should probably be calculating cost for OriginalType to best type and then subtract the cost of CurrentType to best type. 

Dont know what to do about being on the wrong side of the wheel - e.g. you have been terraforming should it become more expensive?
e.g. original type being ocean, and best type terran (so one step counterclock to best). But you are currently at Radiated (so 4 steps clockwise or 5 steps counterclockwise. Cost of reversions was static at roughly 150% of a single step of terraforming. So going the 5 steps (1.5+1.5+1.5+1.5+1) was cheaper than going the 4 steps (4+3+2+1).

also the with the current terraform* buildings, reversions are not cheaper. so, e.g. transforming to an e.g. Terran planet type is the same cost coming a single step clockwise or counter-clockwise. With current content going the 5 steps would still be cheaper (3+2+1+1+1 vs 4+3+2+1)

So here is the plan: i will not implement cheaper reversion. and the automation should still directly go to terran (clockwise) no matter what the best cost is - popular opinion is not always right and this will be a rare case - you still can manually manage if you think it is doing the wrong thing.

Reply:
Actually your calculation for the old formula is wrong, it would have been 5+4+3+2, the cost was DistanceFromOriginal + 1. But then the old Reverse Terraforming did not get cheaper by having Policy Terraforming. I think we better forget about that old formulas.

At the moment, with the targeted version its a trade off. Going the direct way costs a sum of 10xbase and takes 4 time slots, while going the long way costs 8xbase and takes 5 time slots, through most of which you get no intermediate improvement. That's fine, I think. There's also a more extreme case: it could also be 3+4+4+3 vs. 2+1+1+1+2...

Popular terraforming should always go the direct way, going away from the preferred type will not be popular with the planets inhabitants.

As for the formula, the more I ponder on it, the more think we should add an extra interface function for it so we can calculate it in C++, where we can simple use a loop to add the step costs. I think it should be possible to do it in FOCS by manually adding 4 values, each but the first multiplied by 1 or 0 based on whether NextBest(NextBest) != NextBest(NextBest(NextBest)), etc. I guess you get the point. But we'll end up with one hell of a formula.

We could also use that new interface function to simplify the calculations for targeted terraforming, and, if we want, we can then change the cost so that going back isn't cheaper.

Reply:
Actually the sum formula cannot be done in FOCS. Planet has NextBetterPlanetType, but that doesn't allow you to get nextBetter(nextBetter), it's always nextBetter(current).

Reply:
The `NextBetterPlanetType` property uses the species that's on the planet to determine what the next better type is, so a free function just taking a single planet type would not be able to determine that. Instead there'd need to be a `ComplexVariable<PlanetType>` that takes a planet type enum and a string for the species to consider, and that could return another planet type and be chained. This would be similar to the case of `ComplexVariable<int>` that takes two planet types and returns their distance.

Reply:
Well, of course such a function could be added.
My suggestion however is the following:
Add a ComplexInt `TerraformingCostMultiplier planetId=<id> to=<destinationType>`
This would do all those complicated calculation that are in fact not complicated at all when you do them in C++.
E.g. for a Planet with Current = Original = Terran, and to=Inferno it'll return 1+2+3+4=10.
For the way back it would return 3+2+1+1 if we stick to the current rules, but could just as well return 4+3+2+1, if we decide that the way back should cost the same.

Reply:
I see two options currently: one specialised difference function taking three planet types (so distance with a hop in the middle).

Or a difference function with two planet types as parameters, something which is akin to a loop, so you pass the cost function as a valueref and the loop applies the cost function on each step and sums up the values.

Reply:
This is not really resolved (or it needs a followup). @Grummel7 do you think the implementation in this branch (after rebase) is good enough/better than what we have in master?

Reply:
I've just merged it to actually try it out. I found two issues:

1. Build time calculation is wrong, since it is done with integers. E.g. 2 * 90 / 100 = 1. Just remove that part, I do not see why popular terraforming should be faster and cheaper.

2. My original point is still open: Your cost multiplier is N*(N+1)/2*(1.1-N/10), e.g. for 3 steps that give 6*0.8 = 4.8. Doing the first step with targeted terraforming, you pay 1 + 3*0.9 = 3.7.
I thought you were looking for a solution for that after your last comment in February. Btw, if you want a specialised difference function, why not using the one I already suggested? Just give it another name e.g. WeightedPlanetTypeDifference

Reply:
> I've just merged it to actually try it out. I found two issues:
> 
>     1. ...why popular terraforming should be faster and cheaper.

it is the less micro version, it should be obviously better in the normal case

> 
>     2. My original point is still open: Your cost multiplier is N*(N+1)/2*(1.1-N/10), e.g. for 3 steps that give 6_0.8 = 4.8. Doing the first step with targeted terraforming, you pay 1 + 3_0.9 = 3.7.

ah, yes. we could get rid of the complexity by dropping supported use cases. e.g. it only has to be right/cheaper if the planet was not terraformed before (and is more expensive than targeted  terraforming in other cases; by e.g. adding some Max calculation there). in that case using OriginalType and NextBestPlanetType would suffice.

how about 
```
max((PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType))
                             * (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType) / 2 ),
(PlanetTypeDifference from = Target.CurrentType to = Target.NextBestPlanetType))
                             * (PlanetTypeDifference from = Target.CurrentType to = Target.NextBestPlanetType) / 2 ))
```

>        I thought you were looking for a solution for that after your last comment in February. Btw, if you want a specialised difference function, why not using the one I already suggested? Just give it another name e.g. WeightedPlanetTypeDifference

The one you suggested sadly does not suffice. The name is not the problem with your suggested formula. The backend should not know about terraforming costs; that should be scriptable.

i did not find time to implement anything since february, so i expect i wont be able to implement the best solution here in the next half year or so.

Reply:
> > 1. ...why popular terraforming should be faster and cheaper.
> 
> it is the less micro version, it should be obviously better in the normal case

Then I suggest that you multiply by 6/12 before applying the discount. At the moment, due to integer division, two steps take the same time as one, while three and four step take twice the time of one step.

> ah, yes. we could get rid of the complexity by dropping supported use cases. e.g. it only has to be right/cheaper if the planet was not terraformed before (and is more expensive than targeted terraforming in other cases; by e.g. adding some Max calculation there). in that case using OriginalType and NextBestPlanetType would suffice.
> 
> how about
> 
> ```
> max((PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType))
>                              * (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType) / 2 ),
> (PlanetTypeDifference from = Target.CurrentType to = Target.NextBestPlanetType))
>                              * (PlanetTypeDifference from = Target.CurrentType to = Target.NextBestPlanetType) / 2 ))
> ```

That should do, I think.
Using a macro to make it more readable would probably a good idea.

Reply:
implemented the max formula and did some testing; looks ok so far; next step clean up

Reply:
@Grummel7 if you have time i would appreciate if you also test it now and check the two issues are gone